File: rust/tests/debug_channel.rs
use base64::Engine;
use std::fs;
use std::path::PathBuf;
use std::thread;
use std::time::Duration;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let channel_name = args.get(1).unwrap_or(&"debug-channel".to_string()).clone();
    let role = args.get(2).unwrap_or(&"sender".to_string()).clone();

    // Calculate the correct channel directory path
    let channel_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("test")
        .join("interops")
        .join("tmp")
        .join("channels")
        .join(&channel_name);

    println!("=== Rust Channel Debug ===");
    println!("Role: {}", role);
    println!("Channel name: {}", channel_name);
    println!("Channel directory: {:?}", channel_dir);
    println!("CARGO_MANIFEST_DIR: {}", env!("CARGO_MANIFEST_DIR"));
    println!("Current dir: {:?}", std::env::current_dir().unwrap());

    // Ensure channel directory exists
    if !channel_dir.exists() {
        fs::create_dir_all(&channel_dir).expect("Failed to create channel directory");
        println!("Created channel directory");
    } else {
        println!("Channel directory already exists");
    }

    // List current contents
    let entries: Vec<_> = fs::read_dir(&channel_dir)
        .unwrap()
        .filter_map(|e| e.ok())
        .map(|e| e.file_name().to_string_lossy().to_string())
        .collect();

    if entries.is_empty() {
        println!("Current files in channel: (empty)");
    } else {
        println!("Current files in channel: {}", entries.join(", "));
    }

    match role.as_str() {
        "sender" => {
            // Send a test message
            let message_id = format!(
                "{}_{}",
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_millis(),
                0
            );

            let data = b"Hello from Rust!";
            let filename = channel_dir.join(format!("rust_to_ts_{}.msg", message_id));

            let message = serde_json::json!({
                "id": message_id,
                "timestamp": std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_millis() as u64,
                "data": base64::engine::general_purpose::STANDARD.encode(data),
                "length": data.len(),
            });

            fs::write(&filename, serde_json::to_string_pretty(&message).unwrap())
                .expect("Failed to write message");

            println!("\nSent message: {:?}", filename);
            println!(
                "Message content: {}",
                serde_json::to_string_pretty(&message).unwrap()
            );

            // Wait for response
            println!("\nWaiting for response from TypeScript...");
            let mut attempts = 0;
            loop {
                attempts += 1;

                let files: Vec<_> = fs::read_dir(&channel_dir)
                    .unwrap()
                    .filter_map(|e| e.ok())
                    .filter(|e| e.file_name().to_string_lossy().starts_with("ts_to_rust_"))
                    .collect();

                if !files.is_empty() {
                    let first = &files[0];
                    println!("\nReceived response: {:?}", first.file_name());
                    let content = fs::read_to_string(first.path()).unwrap();
                    println!("Response content: {}", content);
                    break;
                }

                if attempts > 30 {
                    println!("\nTimeout waiting for response");
                    std::process::exit(1);
                }

                if attempts % 5 == 0 {
                    println!("Still waiting... ({} seconds)", attempts);
                }

                thread::sleep(Duration::from_secs(1));
            }
        }
        "receiver" => {
            // Wait for a message
            println!("\nWaiting for message from TypeScript...");
            let mut attempts = 0;
            loop {
                attempts += 1;

                let files: Vec<_> = fs::read_dir(&channel_dir)
                    .unwrap()
                    .filter_map(|e| e.ok())
                    .filter(|e| e.file_name().to_string_lossy().starts_with("ts_to_rust_"))
                    .collect();

                if !files.is_empty() {
                    let first = &files[0];
                    println!("\nReceived message: {:?}", first.file_name());
                    let content = fs::read_to_string(first.path()).unwrap();
                    println!("Message content: {}", content);

                    // Parse the message to verify it
                    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();
                    let data_base64 = parsed["data"].as_str().unwrap();
                    let data = base64::engine::general_purpose::STANDARD
                        .decode(data_base64)
                        .unwrap();
                    println!("Decoded data: {}", String::from_utf8_lossy(&data));

                    // Send response
                    let response_id = format!(
                        "{}_{}",
                        std::time::SystemTime::now()
                            .duration_since(std::time::UNIX_EPOCH)
                            .unwrap()
                            .as_millis(),
                        0
                    );

                    let response_data = b"Response from Rust!";
                    let response_filename =
                        channel_dir.join(format!("rust_to_ts_{}.msg", response_id));

                    let response = serde_json::json!({
                        "id": response_id,
                        "timestamp": std::time::SystemTime::now()
                            .duration_since(std::time::UNIX_EPOCH)
                            .unwrap()
                            .as_millis() as u64,
                        "data": base64::engine::general_purpose::STANDARD.encode(response_data),
                        "length": response_data.len(),
                    });

                    fs::write(
                        &response_filename,
                        serde_json::to_string_pretty(&response).unwrap(),
                    )
                    .expect("Failed to write response");

                    println!("\nSent response: {:?}", response_filename);

                    // Clean up the received message
                    fs::remove_file(first.path()).ok();
                    break;
                }

                if attempts > 30 {
                    println!("\nTimeout waiting for message");
                    std::process::exit(1);
                }

                if attempts % 5 == 0 {
                    println!("Still waiting... ({} seconds)", attempts);
                    let current_files: Vec<_> = fs::read_dir(&channel_dir)
                        .unwrap()
                        .filter_map(|e| e.ok())
                        .map(|e| e.file_name().to_string_lossy().to_string())
                        .collect();

                    if current_files.is_empty() {
                        println!("Current files: (empty)");
                    } else {
                        println!("Current files: {}", current_files.join(", "));
                    }
                }

                thread::sleep(Duration::from_secs(1));
            }
        }
        _ => {
            eprintln!("Invalid role. Use 'sender' or 'receiver'");
            std::process::exit(1);
        }
    }
}

File: typescript/test/interops/debug-channel.ts
#!/usr/bin/env ts-node

import * as fs from "fs";
import * as path from "path";
import { Buffer } from "buffer/";

// Debug script to test channel communication between TypeScript and Rust

const channelName = process.argv[2] || "debug-channel";
const role = process.argv[3] || "sender"; // sender or receiver

// Calculate the correct channel directory path
const channelDir = path.join(__dirname, "tmp", "channels", channelName);

console.log("=== Channel Debug Script ===");
console.log(`Role: ${role}`);
console.log(`Channel name: ${channelName}`);
console.log(`Channel directory: ${channelDir}`);
console.log(`__dirname: ${__dirname}`);
console.log(`process.cwd(): ${process.cwd()}`);

// Ensure channel directory exists
if (!fs.existsSync(channelDir)) {
  fs.mkdirSync(channelDir, { recursive: true });
  console.log("Created channel directory");
} else {
  console.log("Channel directory already exists");
}

// List current contents
const files = fs.readdirSync(channelDir);
console.log(`Current files in channel: ${files.length > 0 ? files.join(", ") : "(empty)"}`);

if (role === "sender") {
  // Send a test message
  const messageId = `${Date.now()}_0`;
  const data = Buffer.from("Hello from TypeScript!");
  const filename = path.join(channelDir, `ts_to_rust_${messageId}.msg`);

  const message = {
    id: messageId,
    timestamp: Date.now(),
    data: data.toString("base64"),
    length: data.length,
  };

  fs.writeFileSync(filename, JSON.stringify(message, null, 2));
  console.log(`\nSent message: ${filename}`);
  console.log(`Message content:`, message);

  // Wait and check for response
  console.log("\nWaiting for response from Rust...");
  let attempts = 0;
  const checkInterval = setInterval(() => {
    attempts++;
    const files = fs.readdirSync(channelDir);
    const rustFiles = files.filter(f => f.startsWith("rust_to_ts_"));

    if (rustFiles.length > 0) {
      console.log(`\nReceived response: ${rustFiles[0]}`);
      const content = fs.readFileSync(path.join(channelDir, rustFiles[0]), 'utf-8');
      const response = JSON.parse(content);
      console.log("Response content:", response);
      clearInterval(checkInterval);
      process.exit(0);
    }

    if (attempts > 30) { // 30 seconds timeout
      console.log("\nTimeout waiting for response");
      clearInterval(checkInterval);
      process.exit(1);
    }

    if (attempts % 5 === 0) {
      console.log(`Still waiting... (${attempts} seconds)`);
    }
  }, 1000);

} else if (role === "receiver") {
  // Wait for a message
  console.log("\nWaiting for message from Rust...");
  let attempts = 0;
  const checkInterval = setInterval(() => {
    attempts++;
    const files = fs.readdirSync(channelDir);
    const rustFiles = files.filter(f => f.startsWith("rust_to_ts_"));

    if (rustFiles.length > 0) {
      console.log(`\nReceived message: ${rustFiles[0]}`);
      const content = fs.readFileSync(path.join(channelDir, rustFiles[0]), 'utf-8');
      const message = JSON.parse(content);
      console.log("Message content:", message);

      // Send response
      const responseId = `${Date.now()}_0`;
      const responseData = Buffer.from("Response from TypeScript!");
      const responseFilename = path.join(channelDir, `ts_to_rust_${responseId}.msg`);

      const response = {
        id: responseId,
        timestamp: Date.now(),
        data: responseData.toString("base64"),
        length: responseData.length,
      };

      fs.writeFileSync(responseFilename, JSON.stringify(response, null, 2));
      console.log(`\nSent response: ${responseFilename}`);

      // Clean up the received message
      fs.unlinkSync(path.join(channelDir, rustFiles[0]));

      clearInterval(checkInterval);
      process.exit(0);
    }

    if (attempts > 30) { // 30 seconds timeout
      console.log("\nTimeout waiting for message");
      clearInterval(checkInterval);
      process.exit(1);
    }

    if (attempts % 5 === 0) {
      console.log(`Still waiting... (${attempts} seconds)`);
      const currentFiles = fs.readdirSync(channelDir);
      console.log(`Current files: ${currentFiles.length > 0 ? currentFiles.join(", ") : "(empty)"}`);
    }
  }, 1000);

} else {
  console.error("Invalid role. Use 'sender' or 'receiver'");
  process.exit(1);
}
